### PR TITLE
SALTO-1335 validate types of instances that are not match with types of their Ob…

### PR DESCRIPTION
**Pull request by omrifarkash**
_Release Notes_: 
added recursive validation also for instances value's that aren't matched their parent's objectType

### CODE REVIEW

**Comment by omrifarkash** [comment]
@alonstern  @ori-moisis 
ping
when you have time

File: packages/workspace/test/validator.test.ts [hunk 1]
```diff
@@ -461,6 +461,41 @@ describe('Elements validation', () => {
       )
       expect(errors).toHaveLength(1)
     })
+
+    it('should validated unmatched fields and return unresolved reference error', async () => {
+      const object = new ObjectType({
+        elemID: new ElemID('salto', 'test'),
+        fields: {
+          someFiled: {
```
Comment (alonstern) [inline packages/workspace/test/validator.test.ts:469]: nit: someField

File: packages/workspace/test/validator.test.ts [hunk 2]
```diff
@@ -461,6 +461,41 @@ describe('Elements validation', () => {
...
+            refType: BuiltinTypes.STRING,
+          },
+        },
+      })
+      const objectWithUnMatchedField = new ObjectType({
+        elemID: new ElemID('salto', 'test'),
+        fields: {
+          someFiled: {
```
Comment (alonstern) [inline packages/workspace/test/validator.test.ts:477]: nit: someField

File: packages/workspace/test/validator.test.ts [hunk 3]
```diff
@@ -461,6 +461,41 @@ describe('Elements validation', () => {
       )
       expect(errors).toHaveLength(1)
     })
+
+    it('should validated unmatched fields and return unresolved reference error', async () => {
```
Comment (alonstern) [inline packages/workspace/test/validator.test.ts:465]: I am not sure this test is testing the correct scenario. The scenario you solve in this issue is when validating an instance with a value of a field that does not exist on its type. Here you don't use instances

Comment (omrifarkash) [reply packages/workspace/test/validator.test.ts]: Thanks for reviewing,
Made another push, add some logic for nested fields and add more UTs

File: packages/workspace/src/validator.ts [hunk 4]
```diff
@@ -469,6 +469,28 @@ const validateValue = async (
...
+          { refType: createRefToElmWithValue(BuiltinTypes.UNKNOWN) },
+        ])),
+    })
+    return (await Promise.all(Object.keys(value).map(
+      // eslint-disable-next-line no-use-before-define
+      async k => validateFieldValue(elemID.createNestedID(k),
+        value[k],
+        (await objType.fields[k]?.getType(elementsSource)) ?? BuiltinTypes.UNKNOWN,
+        objType.fields[k]?.annotations ?? {},
+        elementsSource)
```
Comment (alonstern) [inline packages/workspace/src/validator.ts:490]:
```suggestion
    return (await Promise.all(Object.keys(value).map(
      // eslint-disable-next-line no-use-before-define
      async k => validateFieldValue(elemID.createNestedID(k),
        value[k],
        BuiltinTypes.UNKNOWN,
        {},
        elementsSource)
```

File: packages/workspace/test/validator.test.ts [hunk 5]
```diff
@@ -1905,6 +1870,103 @@ describe('Elements validation', () => {
...
+      const someType = new ObjectType({
+        elemID: new ElemID('salto', 'test'),
+        fields: {
+          someField: {
+            refType: createRefToElmWithValue(BuiltinTypes.STRING),
+          },
+        },
+      })
+
+      it('should validated unmatched field with legal value with no errors', async () => {
```
Comment (alonstern) [inline packages/workspace/test/validator.test.ts:1884]: nit : `should validate unmatched field with legal value with no errors`

File: packages/workspace/test/validator.test.ts [hunk 6]
```diff
@@ -1905,6 +1870,103 @@ describe('Elements validation', () => {
           .toEqual(unresolvedRefInAnnoInst.elemID.createNestedID(CORE_ANNOTATIONS.PARENT, '1'))
       })
     })
+
+    describe('validateFieldsThatAreUnMatchedToObjectType', () => {
```
Comment (alonstern) [inline packages/workspace/test/validator.test.ts:1874]: nit: `validate fields that are unmatched to ObjectType`

**Review by alonstern — APPROVED** [review]
Approving with some nits
Once you fix a comment, please mark the Resolve conversion button (unless you add a comment you wish to reviewer to see).

**Pull request merged**